microsoft_teams/plugin.spec.yaml

```diff
@@ -266,18 +276,28 @@ actions:
         type: string
         required: true
         order: 1
+        example: InsightConnect Team
       channel_name:
         title: Channel Name
         description: Channel
         type: string
         required: true
         order: 2
+        example: InsightConnect Channel
+      thread_id:
+        title: Thread ID
+        description: To post in a thread, select parent message ID
+        type: string
+        required: true
```

Comment:
@dmandyna-r7 Does this need to be required? We should allow users to choose threads if they want to.

If it's optional, then this can be a minor version bump.

Reply:
Agree, that was oversight on my side - I've changed the code so the field is no longer required, and bumped minor version number instead of major.

Reply:
## Assessment
### Run

<details>

```
{
  "body": {
    "log": "Time Now: 1596113911.596647\nTime Ago: 0\nRefreshing auth token\nUpdating Auth Token...\nGetting token from: https://login.microsoftonline.com/5c824599-dc8c-4d31-96fb-3b886d4f8f10/oauth2/token\nAuthentication was successful, token is: ******************FGL8g\nDetected Permissions: Agreement.Read.All Agreement.ReadWrite.All AgreementAcceptance.Read AgreementAcceptance.Read.All AppCatalog.ReadWrite.All AuditLog.Read.All Bookings.Manage.All Bookings.Read.All Bookings.ReadWrite.All BookingsAppointment.ReadWrite.All Calendars.Read Calendars.Read.Shared Calendars.ReadWrite Calendars.ReadWrite.Shared Contacts.Read Contacts.Read.Shared Contacts.ReadWrite Contacts.ReadWrite.Shared Device.Command Device.Read DeviceManagementApps.Read.All DeviceManagementApps.ReadWrite.All DeviceManagementConfiguration.Read.All DeviceManagementConfiguration.ReadWrite.All DeviceManagementManagedDevices.PrivilegedOperations.All DeviceManagementManagedDevices.Read.All DeviceManagementManagedDevices.ReadWrite.All DeviceManagementRBAC.Read.All DeviceManagementRBAC.ReadWrite.All DeviceManagementServiceConfig.Read.All DeviceManagementServiceConfig.ReadWrite.All Directory.AccessAsUser.All Directory.ReadWrite.All EAS.AccessAsUser.All EduAdministration.Read EduAdministration.ReadWrite EduAssignments.Read EduAssignments.ReadBasic EduAssignments.ReadWrite EduAssignments.ReadWriteBasic EduRoster.Read EduRoster.ReadBasic EduRoster.ReadWrite email Files.Read Files.Read.All Files.Read.Selected Files.ReadWrite Files.ReadWrite.All Files.ReadWrite.AppFolder Files.ReadWrite.Selected Financials.ReadWrite.All Group.Read.All Group.ReadWrite.All IdentityProvider.Read.All IdentityProvider.ReadWrite.All IdentityRiskEvent.Read.All Mail.Read Mail.Read.Shared Mail.ReadWrite Mail.ReadWrite.Shared Mail.Send Mail.Send.Shared MailboxSettings.Read MailboxSettings.ReadWrite Member.Read.Hidden Notes.Create Notes.Read Notes.Read.All Notes.ReadWrite Notes.ReadWrite.All Notes.ReadWrite.CreatedByApp offline_access openid People.Read People.Read.All PrivilegedAccess.ReadWrite.AzureAD PrivilegedAccess.ReadWrite.AzureResources profile Reports.Read.All SecurityEvents.Read.All SecurityEvents.ReadWrite.All Sites.FullControl.All Sites.Manage.All Sites.Read.All Sites.ReadWrite.All Subscription.Read.All Tasks.Read Tasks.Read.Shared Tasks.ReadWrite Tasks.ReadWrite.Shared User.Export.All User.Invite.All User.Read.All User.ReadWrite.All UserActivity.ReadWrite.CreatedByApp UserTimelineActivity.Write.CreatedByApp\nrapid7/Microsoft Teams:3.0.0. Step name: send_html_message\nTime Now: 1596113912.9509592\nTime Ago: 1596113912.9487472\nToken is valid, not refreshing.\nTeam name: test_group_delete_me\nChecking team: test_group_delete_me\nTime Now: 1596113913.4302456\nTime Ago: 1596113912.9487472\nToken is valid, not refreshing.\nChannel name: General\nChecking channel: General\nSending message to: https://graph.microsoft.com/beta/teams/46d3519a-8061-418e-ada9-f7aaa4598e14/channels/19:73fb8054746f43a8a3eeacfb68d15e1c@thread.tacv2/messages/1595889908700/replies\nTime Now: 1596113914.8153303\nTime Ago: 1596113912.9487472\nToken is valid, not refreshing.\n",
    "meta": {},
    "output": {
      "message": {
        "@odata.context": "https://graph.microsoft.com/beta/$metadata#teams('46d3519a-8061-418e-ada9-f7aaa4598e14')/channels('19%!A(MISSING)73fb8054746f43a8a3eeacfb68d15e1c%!t(MISSING)hread.tacv2')/messages('1595889908700')/replies/$entity",
        "attachments": [],
        "body": {
          "content": "\u003ch3\u003eThis is HTML reply in a thread!\u003c/h3\u003e",
          "contentType": "html"
        },
        "channelIdentity": {
          "channelId": "19:73fb8054746f43a8a3eeacfb68d15e1c@thread.tacv2",
          "teamId": "46d3519a-8061-418e-ada9-f7aaa4598e14"
        },
        "createdDateTime": "2020-07-30T12:58:36.919Z",
        "etag": "1596113916919",
        "from": {
          "user": {
            "displayName": "Jon Schipp",
            "id": "ac785ffe-530a-45a1-bbf4-e275457e464b",
            "userIdentityType": "aadUser"
          }
        },
        "id": "1596113916919",
        "importance": "normal",
        "locale": "en-us",
        "mentions": [],
        "messageType": "message",
        "reactions": [],
        "replyToId": "1595889908700",
        "webUrl": "https://teams.microsoft.com/l/message/19%!A(MISSING)73fb8054746f43a8a3eeacfb68d15e1c%!t(MISSING)hread.tacv2/1596113916919?groupId=46d3519a-8061-418e-ada9-f7aaa4598e14\u0026tenantId=5c824599-dc8c-4d31-96fb-3b886d4f8f10\u0026createdTime=1596113916919\u0026parentMessageId=1595889908700"
      }
    },
    "status": "ok"
  },
  "type": "action_event",
  "version": "v1"
}

```

<summary>
docker run --rm -i rapid7/microsoft_teams:2.1.0 --debug run < tests/send_html_message.json
</summary>
</details>

<details>

```
{
  "body": {
    "log": "Time Now: 1596113920.4204874\nTime Ago: 0\nRefreshing auth token\nUpdating Auth Token...\nGetting token from: https://login.microsoftonline.com/5c824599-dc8c-4d31-96fb-3b886d4f8f10/oauth2/token\nAuthentication was successful, token is: ******************kOPyg\nDetected Permissions: Agreement.Read.All Agreement.ReadWrite.All AgreementAcceptance.Read AgreementAcceptance.Read.All AppCatalog.ReadWrite.All AuditLog.Read.All Bookings.Manage.All Bookings.Read.All Bookings.ReadWrite.All BookingsAppointment.ReadWrite.All Calendars.Read Calendars.Read.Shared Calendars.ReadWrite Calendars.ReadWrite.Shared Contacts.Read Contacts.Read.Shared Contacts.ReadWrite Contacts.ReadWrite.Shared Device.Command Device.Read DeviceManagementApps.Read.All DeviceManagementApps.ReadWrite.All DeviceManagementConfiguration.Read.All DeviceManagementConfiguration.ReadWrite.All DeviceManagementManagedDevices.PrivilegedOperations.All DeviceManagementManagedDevices.Read.All DeviceManagementManagedDevices.ReadWrite.All DeviceManagementRBAC.Read.All DeviceManagementRBAC.ReadWrite.All DeviceManagementServiceConfig.Read.All DeviceManagementServiceConfig.ReadWrite.All Directory.AccessAsUser.All Directory.ReadWrite.All EAS.AccessAsUser.All EduAdministration.Read EduAdministration.ReadWrite EduAssignments.Read EduAssignments.ReadBasic EduAssignments.ReadWrite EduAssignments.ReadWriteBasic EduRoster.Read EduRoster.ReadBasic EduRoster.ReadWrite email Files.Read Files.Read.All Files.Read.Selected Files.ReadWrite Files.ReadWrite.All Files.ReadWrite.AppFolder Files.ReadWrite.Selected Financials.ReadWrite.All Group.Read.All Group.ReadWrite.All IdentityProvider.Read.All IdentityProvider.ReadWrite.All IdentityRiskEvent.Read.All Mail.Read Mail.Read.Shared Mail.ReadWrite Mail.ReadWrite.Shared Mail.Send Mail.Send.Shared MailboxSettings.Read MailboxSettings.ReadWrite Member.Read.Hidden Notes.Create Notes.Read Notes.Read.All Notes.ReadWrite Notes.ReadWrite.All Notes.ReadWrite.CreatedByApp offline_access openid People.Read People.Read.All PrivilegedAccess.ReadWrite.AzureAD PrivilegedAccess.ReadWrite.AzureResources profile Reports.Read.All SecurityEvents.Read.All SecurityEvents.ReadWrite.All Sites.FullControl.All Sites.Manage.All Sites.Read.All Sites.ReadWrite.All Subscription.Read.All Tasks.Read Tasks.Read.Shared Tasks.ReadWrite Tasks.ReadWrite.Shared User.Export.All User.Invite.All User.Read.All User.ReadWrite.All UserActivity.ReadWrite.CreatedByApp UserTimelineActivity.Write.CreatedByApp\nrapid7/Microsoft Teams:3.0.0. Step name: send_html_message\nTime Now: 1596113921.5254905\nTime Ago: 1596113921.5230486\nToken is valid, not refreshing.\nTeam name: test_group_delete_me\nChecking team: test_group_delete_me\nTime Now: 1596113921.8240914\nTime Ago: 1596113921.5230486\nToken is valid, not refreshing.\nChannel name: General\nChecking channel: General\nSending message to: https://graph.microsoft.com/beta/teams/46d3519a-8061-418e-ada9-f7aaa4598e14/channels/19:73fb8054746f43a8a3eeacfb68d15e1c@thread.tacv2/messages\nTime Now: 1596113922.8329384\nTime Ago: 1596113921.5230486\nToken is valid, not refreshing.\n",
    "meta": {},
    "output": {
      "message": {
        "@odata.context": "https://graph.microsoft.com/beta/$metadata#teams('46d3519a-8061-418e-ada9-f7aaa4598e14')/channels('19%!A(MISSING)73fb8054746f43a8a3eeacfb68d15e1c%!t(MISSING)hread.tacv2')/messages/$entity",
        "attachments": [],
        "body": {
          "content": "\u003ch3\u003eThis is HTML message!\u003c/h3\u003e",
          "contentType": "html"
        },
        "channelIdentity": {
          "channelId": "19:73fb8054746f43a8a3eeacfb68d15e1c@thread.tacv2",
          "teamId": "46d3519a-8061-418e-ada9-f7aaa4598e14"
        },
        "createdDateTime": "2020-07-30T12:58:44.317Z",
        "etag": "1596113924317",
        "from": {
          "user": {
            "displayName": "Jon Schipp",
            "id": "ac785ffe-530a-45a1-bbf4-e275457e464b",
            "userIdentityType": "aadUser"
          }
        },
        "id": "1596113924317",
        "importance": "normal",
        "locale": "en-us",
        "mentions": [],
        "messageType": "message",
        "reactions": [],
        "webUrl": "https://teams.microsoft.com/l/message/19%!A(MISSING)73fb8054746f43a8a3eeacfb68d15e1c%!t(MISSING)hread.tacv2/1596113924317?groupId=46d3519a-8061-418e-ada9-f7aaa4598e14\u0026tenantId=5c824599-dc8c-4d31-96fb-3b886d4f8f10\u0026createdTime=1596113924317\u0026parentMessageId=1596113924317"
      }
    },
    "status": "ok"
  },
  "type": "action_event",
  "version": "v1"
}

```

<summary>
docker run --rm -i rapid7/microsoft_teams:2.1.0 --debug run < tests/send_html_message_no_thread.json
</summary>
</details>

<details>

```
{
  "body": {
    "log": "Time Now: 1596113927.749738\nTime Ago: 0\nRefreshing auth token\nUpdating Auth Token...\nGetting token from: https://login.microsoftonline.com/5c824599-dc8c-4d31-96fb-3b886d4f8f10/oauth2/token\nAuthentication was successful, token is: ******************gIY_w\nDetected Permissions: Agreement.Read.All Agreement.ReadWrite.All AgreementAcceptance.Read AgreementAcceptance.Read.All AppCatalog.ReadWrite.All AuditLog.Read.All Bookings.Manage.All Bookings.Read.All Bookings.ReadWrite.All BookingsAppointment.ReadWrite.All Calendars.Read Calendars.Read.Shared Calendars.ReadWrite Calendars.ReadWrite.Shared Contacts.Read Contacts.Read.Shared Contacts.ReadWrite Contacts.ReadWrite.Shared Device.Command Device.Read DeviceManagementApps.Read.All DeviceManagementApps.ReadWrite.All DeviceManagementConfiguration.Read.All DeviceManagementConfiguration.ReadWrite.All DeviceManagementManagedDevices.PrivilegedOperations.All DeviceManagementManagedDevices.Read.All DeviceManagementManagedDevices.ReadWrite.All DeviceManagementRBAC.Read.All DeviceManagementRBAC.ReadWrite.All DeviceManagementServiceConfig.Read.All DeviceManagementServiceConfig.ReadWrite.All Directory.AccessAsUser.All Directory.ReadWrite.All EAS.AccessAsUser.All EduAdministration.Read EduAdministration.ReadWrite EduAssignments.Read EduAssignments.ReadBasic EduAssignments.ReadWrite EduAssignments.ReadWriteBasic EduRoster.Read EduRoster.ReadBasic EduRoster.ReadWrite email Files.Read Files.Read.All Files.Read.Selected Files.ReadWrite Files.ReadWrite.All Files.ReadWrite.AppFolder Files.ReadWrite.Selected Financials.ReadWrite.All Group.Read.All Group.ReadWrite.All IdentityProvider.Read.All IdentityProvider.ReadWrite.All IdentityRiskEvent.Read.All Mail.Read Mail.Read.Shared Mail.ReadWrite Mail.ReadWrite.Shared Mail.Send Mail.Send.Shared MailboxSettings.Read MailboxSettings.ReadWrite Member.Read.Hidden Notes.Create Notes.Read Notes.Read.All Notes.ReadWrite Notes.ReadWrite.All Notes.ReadWrite.CreatedByApp offline_access openid People.Read People.Read.All PrivilegedAccess.ReadWrite.AzureAD PrivilegedAccess.ReadWrite.AzureResources profile Reports.Read.All SecurityEvents.Read.All SecurityEvents.ReadWrite.All Sites.FullControl.All Sites.Manage.All Sites.Read.All Sites.ReadWrite.All Subscription.Read.All Tasks.Read Tasks.Read.Shared Tasks.ReadWrite Tasks.ReadWrite.Shared User.Export.All User.Invite.All User.Read.All User.ReadWrite.All UserActivity.ReadWrite.CreatedByApp UserTimelineActivity.Write.CreatedByApp\nrapid7/Microsoft Teams:3.0.0. Step name: send_message\nTime Now: 1596113928.3400035\nTime Ago: 1596113928.336709\nToken is valid, not refreshing.\nTeam name: test_group_delete_me\nChecking team: test_group_delete_me\nTime Now: 1596113928.6643672\nTime Ago: 1596113928.336709\nToken is valid, not refreshing.\nChannel name: General\nChecking channel: General\nSending message to: https://graph.microsoft.com/beta/teams/46d3519a-8061-418e-ada9-f7aaa4598e14/channels/19:73fb8054746f43a8a3eeacfb68d15e1c@thread.tacv2/messages/1595889908700/replies\nTime Now: 1596113929.6936533\nTime Ago: 1596113928.336709\nToken is valid, not refreshing.\n",
    "meta": {},
    "output": {
      "message": {
        "@odata.context": "https://graph.microsoft.com/beta/$metadata#teams('46d3519a-8061-418e-ada9-f7aaa4598e14')/channels('19%!A(MISSING)73fb8054746f43a8a3eeacfb68d15e1c%!t(MISSING)hread.tacv2')/messages('1595889908700')/replies/$entity",
        "attachments": [],
        "body": {
          "content": "This is text reply in a thread!",
          "contentType": "text"
        },
        "channelIdentity": {
          "channelId": "19:73fb8054746f43a8a3eeacfb68d15e1c@thread.tacv2",
          "teamId": "46d3519a-8061-418e-ada9-f7aaa4598e14"
        },
        "createdDateTime": "2020-07-30T12:58:51.195Z",
        "etag": "1596113931195",
        "from": {
          "user": {
            "displayName": "Jon Schipp",
            "id": "ac785ffe-530a-45a1-bbf4-e275457e464b",
            "userIdentityType": "aadUser"
          }
        },
        "id": "1596113931195",
        "importance": "normal",
        "locale": "en-us",
        "mentions": [],
        "messageType": "message",
        "reactions": [],
        "replyToId": "1595889908700",
        "webUrl": "https://teams.microsoft.com/l/message/19%!A(MISSING)73fb8054746f43a8a3eeacfb68d15e1c%!t(MISSING)hread.tacv2/1596113931195?groupId=46d3519a-8061-418e-ada9-f7aaa4598e14\u0026tenantId=5c824599-dc8c-4d31-96fb-3b886d4f8f10\u0026createdTime=1596113931195\u0026parentMessageId=1595889908700"
      }
    },
    "status": "ok"
  },
  "type": "action_event",
  "version": "v1"
}

```

<summary>
docker run --rm -i rapid7/microsoft_teams:2.1.0 --debug run < tests/send_message.json
</summary>
</details>